src/Datadog.Trace.ClrProfiler.Native/clr_helpers.cpp

```diff
@@ -682,4 +683,212 @@ bool TryParseSignatureTypes(const ComPtr<IMetaDataImport2>& metadata_import,
+    case ELEMENT_TYPE_R4:            // 0x0c  // System.Single (struct)
+    case ELEMENT_TYPE_R8:            // 0x0d  // System.Double (struct)
+      return true;
+
+    // The following are primitives that are not valuetypes
+    case ELEMENT_TYPE_STRING:        // 0x0e  // System.String
+    case ELEMENT_TYPE_PTR:           // 0x0f  // No managed type
+    case ELEMENT_TYPE_BYREF:         // 0x10  // No managed type
+      return false;
+
+    case ELEMENT_TYPE_VALUETYPE:     // 0x11  // Valuetype. Return true
+      return true;
+
+    case ELEMENT_TYPE_CLASS:         // 0x12  // Reference type. Return false
+    case ELEMENT_TYPE_VAR:           // 0x13  // Depends on the runtime type. Return false
+    case ELEMENT_TYPE_ARRAY:         // 0x14  // Array reference. Return false
+    case ELEMENT_TYPE_GENERICINST:   // 0X15  // Depends on the defining type. Return false
+      return false;
+
+    case ELEMENT_TYPE_TYPEDBYREF:    // 0X16  // System.TypedReference (struct)
+    case ELEMENT_TYPE_I:             // 0x18  // System.IntPtr (struct)
+    case ELEMENT_TYPE_U:             // 0x19  // System.UIntPtr (struct)
+      return true;
+
+    case ELEMENT_TYPE_FNPTR:         // 0x1b  // No managed type
+    case ELEMENT_TYPE_OBJECT:        // 0x1c  // Shortcut for System.Object, return false
+    case ELEMENT_TYPE_SZARRAY:       // 0x1d  // Shortcut for single dimension zero lower bound array. return false
+    case ELEMENT_TYPE_MVAR:          // 0x1e  // Depends on the runtime type. Return false
+      return false;
+
+    default:
+      Warn("[trace::ElementTypeIsAlwaysValueType] UNHANDLED CASE: unexpected CorElementType found: ", element_type);
+      return false;
+  }
+}
+
+bool ReturnTypeIsValueTypeOrGeneric(const ComPtr<IMetaDataImport2>& metadata_import,
+                      const ComPtr<IMetaDataEmit2>& metadata_emit,
+                      const mdToken targetFunctionToken,
+                      const MethodSignature targetFunctionSignature,
+                      mdToken* ret_type_token) {
+
+  // MethodDefSig Format: [[HASTHIS] [EXPLICITTHIS]] (DEFAULT|VARARG|GENERIC GenParamCount) ParamCount RetType Param* [SENTINEL Param+]
+  const auto generic_count = targetFunctionSignature.NumberOfTypeArguments();
+  size_t method_def_sig_index = generic_count == 0 ? 2 : 3;  // Initialize the index to point to RetType
+  auto ret_type_byte = targetFunctionSignature.data[method_def_sig_index];
+  const auto ret_type = CorElementType(ret_type_byte);
+
+  switch (ret_type) {
+    case ELEMENT_TYPE_VOID:
+      // No object is returned, so return false.
+      return false;
+
+    case ELEMENT_TYPE_GENERICINST:
+      // Format: GENERICINST (CLASS | VALUETYPE) TypeDefOrRefEncoded GenArgCount Type *
+      // Example: Task<HttpResponseMessage>. Return true if the type is a VALUETYPE
+      method_def_sig_index++;
+      return targetFunctionSignature.data[method_def_sig_index] == ELEMENT_TYPE_VALUETYPE;
```

Comment:
I just realized that here I am returning true but didn't actually set `ret_type_token` which is the type token to unbox to . I'll need to improve my tests and also figure this part out.

Reply:
Commit 6b381fbb42efc8e9892015da508267fa584730a2 includes all the changes necessary to return the correct TypeRef for the value type.